crates/oxc_linter/src/rules/unicorn/no_process_exit.rs

```diff
@@ -0,0 +1,188 @@
+use oxc_ast::AstKind;
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{ast_util::is_method_call, context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint-plugin-unicorn(no-process-exit): Disallow `process.exit()`.")]
+#[diagnostic(
+    severity(warning),
+    help("Only use `process.exit()` in CLI apps. Throw an error instead.")
+)]
+struct NoProcessExitDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoProcessExit;
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Disallow `process.exit()`.
+    ///
+    /// ### Why is this bad?
+    /// Only use `process.exit()` in CLI apps. Throw an error instead.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// // Bad
+    /// if (problem) process.exit(1);
+    ///
+    /// // Good
+    /// if (problem) throw new Error("message");
+    ///
+    /// #! /usr/bin/env node
+    /// if (problem) process.exit(1);
+    /// ```
+    NoProcessExit,
+    restriction,
+);
+
+impl Rule for NoProcessExit {
+    fn run_once<'a>(&self, ctx: &LintContext<'_>) {
+        // Allowed in shebang
+        let Some(root) = ctx.nodes().iter().next() else { return };
+        let AstKind::Program(program) = root.kind() else { return };
+        if program.hashbang.is_some() {
+            return;
+        }
+
+        for node in ctx.nodes().iter() {
+            if let AstKind::CallExpression(expr) = node.kind() {
+                if is_method_call(expr, Some(&["process"]), Some(&["exit"]), None, None) {
+                    if is_inside_process_event_handler(ctx, node) || is_worker_threads_imported(ctx)
+                    {
+                        return;
+                    }
+
+                    ctx.diagnostic(NoProcessExitDiagnostic(expr.span));
+                }
+            }
+        }
+    }
```

Comment:
sorry don't think this should be `run_once` as we are going to iter over every node on L53.

i think it should be something like:

```rs
impl Rule for NoProcessExit {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        if let AstKind::CallExpression(expr) = node.kind() {
            if is_method_call(expr, Some(&["process"]), Some(&["exit"]), None, None) {
                let Some(root) = ctx.nodes().iter().next() else { return };
                let AstKind::Program(program) = root.kind() else { return };
                if program.hashbang.is_some() {
                    return;
                }
                if is_inside_process_event_handler(ctx, node) || is_worker_threads_imported(ctx) {
                    return;
                }

                ctx.diagnostic(NoProcessExitDiagnostic(expr.span));
            }
        }
    }
}
```

That way, we avoid 2x passes over the node (one by the main linter, one here in `for node in ctx.nodes().iter()`, while still checking all the cases.

What do you think?

Reply:
I switched from `run` to `run_once` because,
- hashbang only appears once
- So, wouldn't it be inefficient for each node to check it every time?

But as you say, Linter calls `run` even if it is not implemented, it seems better to check it within `run`.

- - -

👉🏻 [76b5070](https://github.com/oxc-project/oxc/pull/2410/commits/76b5070b847e910a7421d5e1b3e8a8367a582d37)

Reply:
thanks!